File: reorder-list/reorder-list.cs
            return;
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast!=null && fast.next!=null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        Console.Out.WriteLine("slow is {0}",slow.val);
        Console.Out.WriteLine(fast);
​
        
        fast = reverse(slow.next);
        slow.next=null;
        slow = head;
        
        while(fast!=null)
        {
            ListNode temp = slow.next;
            slow.next = fast;
            fast = fast.next;
            slow.next.next = temp;
            slow = temp;
        }
        
        
        
    }
    
    private ListNode reverse(ListNode head)
    {
        if(head==null || head.next==null)
        {
            return head;
        }
        ListNode reversed = reverse(head.next);
        head.next.next= head;
        head.next=null;
        return reversed;
    }
}

File: java/ReorderList.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        //find the mid point
        ListNode slow = head, fast = head;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }

        //reverse the second half the list
        ListNode prev=null;
        ListNode curr = slow, tmp;
        while (curr!=null){
            tmp = curr.next;
            curr.next = prev;
            prev=curr;
            curr= tmp;
        }


        //zigzag the pointers
        //connect them both
        ListNode first = head, second = prev;
        while (second.next != null) {
            tmp = first.next;
            first.next = second;
            first = tmp;

            tmp = second.next;
            second.next = first;
            second = tmp;
        }


    }
}